sql/log_event.cc

```diff
@@ -14521,3 +14560,421 @@ size_t my_strmov_quoted_identifier_helper(int q, char *buffer,
+                  ->prev_event != nullptr);
+  DBUG_ASSERT(ev->begin_event() != nullptr);
+  DBUG_ASSERT(!ev->begin_event()->whole_group_scheduled);
+
+  ev->is_end_event = true;
+  ev->begin_event()->whole_group_scheduled = true;
+
+  DBUG_VOID_RETURN;
+}
+
+/**
+  Callback to deallocate a given key buffer. This is passed as a parameter
+  while creating shared pointers to the key_buf.
+*/
+void key_dealloc_cb(uchar *key_buf) { my_free(key_buf); }
+
+bool Rows_log_event::parse_keys(Relay_log_info *rli,
+                                std::shared_ptr<Log_event_wrapper> &,
+                                TABLE *table,
+                                std::deque<Dependency_key> &keys) {
+  DBUG_ENTER("Rows_log_event::parse_keys");
+  std::vector<KEY *> key_infos;
+
+  // Find PK and all unique keys
+  uint key = 0;
+  KEY *keyinfo = table->key_info;
+  bool found_pk = false;
+  for (; key < table->s->keys; key++, keyinfo++) {
+    // Skip non-unique keys
+    if (!(keyinfo->flags & HA_NOSAME)) continue;
+
+    // NOTE: We expect the before image to contain all the unique keys
+    // (including PK), if any key is not present we fail fast and process the
+    // transaction in sync mode (see: @Rows_log_event::prepare_dep)
+    if (!are_all_columns_signaled_for_key(keyinfo, &m_cols)) DBUG_RETURN(false);
+
+    if (key == table->s->primary_key) found_pk = true;
+    key_infos.push_back(&table->key_info[key]);
+  }
+
+  // Case: We cannot function without a PK, abort!
+  if (!found_pk) DBUG_RETURN(false);
+
+  // Figure out type conversions
+  TABLE *conv_table = nullptr;
+  if (!rli->tables_to_lock->m_tabledef.compatible_with(
+          thd, rli, rli->tables_to_lock->table, &conv_table)) {
+    sql_print_error(
+        "Unable to check type compatibility at pos %s:%llu, "
+        "syncing group",
+        rli->get_rpl_log_name(), common_header->log_pos);
+    clear_all_errors(rli->info_thd, rli);
+    DBUG_RETURN(false);
+  }
+  rli->tables_to_lock->m_conv_table = conv_table;
+
+  uchar *curr_row = nullptr, *key_buf = nullptr, *curr_row_end = nullptr;
+  uint i = 0;
+  for (i = 0, curr_row = m_rows_buf; curr_row != m_rows_end;
+       curr_row = curr_row_end, i++) {
+    MY_BITMAP *cols = &m_cols;  // init with before image bitmap
+
+    enum_row_image_type row_image_type =
+        get_general_type_code() == binary_log::WRITE_ROWS_EVENT
+            ? enum_row_image_type::WRITE_AI
+            : get_general_type_code() == binary_log::DELETE_ROWS_EVENT
+                  ? enum_row_image_type::DELETE_BI
+                  : enum_row_image_type::UPDATE_BI;
+    bool has_value_options =
+        (get_type_code() == binary_log::PARTIAL_UPDATE_ROWS_EVENT);
+
+    // This is needed only to satisfy some debug asserts, not strictly required
+    // for key parsing because upack_row takes cols as an argument instead of
+    // relying on read and write sets
+    bitmap_set_all(table->read_set);
+    bitmap_set_all(table->write_set);
```

Comment:
Can these lines be moved outside the loop?

Reply:
It seems to work fine outside the loop. For example `rpl_mts_dependency_type_conv` fails without `bitmap_set_all` but works with `bitmap_set_all` moved outside the loop.